File: services/payment-service/src/services/saga.service.ts
import type { PoolClient } from 'pg';
import { config } from '../config';
import {
  MAX_SETTLE_ATTEMPTS,
  backoffMs,
  canCompensate,
  canSettle,
  isExhausted,
  moveFunds,
  shouldSimulateFailure,
  type PaymentStatus,
} from '../domain/payment';
import { toEventBody, type PaymentRow } from '../models/payment.model';
import * as accounts from '../repositories/account.repository';
import * as ledger from '../repositories/ledger.repository';
import * as outbox from '../repositories/outbox.repository';
import * as payments from '../repositories/payment.repository';

const { clearingId } = config.systemAccounts;

/**
 * Leg 2 of the saga: move the held funds from clearing to the receiver.
 *
 * Runs in its own transaction, some time after leg 1 - that separation is the
 * entire reason a payment can get stuck, and the reason a compensating action
 * has to exist at all.
 */
export async function settle(
  client: PoolClient,
  row: PaymentRow,
): Promise<PaymentStatus> {
  if (!canSettle(row.status)) return row.status;

  /** Out of retries: the money stays in clearing, owed back to the sender. */
  const strand = async (reason: string): Promise<PaymentStatus> => {
    const updatedAt = await payments.markStranded(
      client,
      row.id,
      reason,
      row.attempts + 1,
      config.saga.compensateDelayMs,
    );
    await outbox.enqueue(client, 'payment.stuck', {
      ...toEventBody(row, updatedAt),
      failureReason: reason,
      attempts: row.attempts + 1,
    });
    return 'AWAITING_REFUND';
  };

  /**
   * A failure is not automatically fatal. Most things that break between two
   * services break briefly, so try again with a backoff and only give the
   * money back once the attempts are used up. Compensating on the first
   * hiccup would unwind perfectly good payments.
   */
  const retryOrStrand = async (reason: string): Promise<PaymentStatus> => {
    const attempts = row.attempts + 1;
    if (isExhausted(attempts)) return strand(reason);

    const delay = backoffMs(attempts, { jitter: Math.random() });
    const updatedAt = await payments.scheduleRetry(client, row.id, attempts, reason, delay);
    await outbox.enqueue(client, 'payment.settlement_retrying', {
      ...toEventBody(row, updatedAt),
      failureReason: reason,
      attempts,
      maxAttempts: MAX_SETTLE_ATTEMPTS,
      retryInMs: delay,
    });
    return 'PROCESSING';
  };

  if (shouldSimulateFailure(row.simulate_mode, row.attempts)) {
    return retryOrStrand('SETTLEMENT_FAILED_SIMULATED');
  }

  const locked = await accounts.lockMany(client, [clearingId, row.to_account_id]);
  const clearing = locked.get(clearingId);
  const receiver = locked.get(row.to_account_id);
  if (!clearing || !receiver) return retryOrStrand('RECEIVER_UNAVAILABLE');

  const move = moveFunds(clearing, receiver, row.amount_cents);
  if (!move.ok) return retryOrStrand(move.failureReason);

  await accounts.updateBalance(client, clearing.id, move.fromBalanceCents);
  await accounts.updateBalance(client, receiver.id, move.toBalanceCents);
  await ledger.postJournal(client, row.id, 'SETTLE', move.entries);

  const updatedAt = await payments.markCompleted(client, row.id, row.attempts + 1);
  await outbox.enqueue(client, 'payment.completed', {
    ...toEventBody(row, updatedAt),
    attempts: row.attempts + 1,
  });
  return 'COMPLETED';
}

/**
 * The compensating action: return stranded funds from clearing to the sender.
 *
 * Shared by the automatic worker, the manual refund endpoint and a rejected
 * review, so all three take exactly the same path and cannot drift apart.
 */
export async function compensate(
  client: PoolClient,
  row: PaymentRow,
  reason?: string,
): Promise<PaymentStatus> {
  if (!canCompensate(row.status)) return row.status;

  const locked = await accounts.lockMany(client, [clearingId, row.from_account_id]);
  const clearing = locked.get(clearingId)!;
  const sender = locked.get(row.from_account_id)!;

  const move = moveFunds(clearing, sender, row.amount_cents);
  if (!move.ok) {
    // Clearing not holding the money means the ledger has been tampered with.
    // Refuse loudly rather than invent a balance.
    throw new Error(`cannot compensate ${row.id}: clearing account ${move.failureReason}`);
  }

  await accounts.updateBalance(client, clearing.id, move.fromBalanceCents);
  await accounts.updateBalance(client, sender.id, move.toBalanceCents);
  await ledger.postJournal(client, row.id, 'COMPENSATE', move.entries);

  const updatedAt = await payments.markRefunded(client, row.id, reason);
  await outbox.enqueue(client, 'payment.refunded', {
    ...toEventBody(row, updatedAt),
    ...(reason ? { failureReason: reason } : {}),
  });
  return 'REFUNDED';
}

File: services/payment-service/src/main/java/com/ledgerflow/payment/services/SagaService.java
package com.ledgerflow.payment.services;

import com.ledgerflow.payment.config.Config;
import com.ledgerflow.payment.domain.Leg;
import com.ledgerflow.payment.domain.PaymentStatus;
import com.ledgerflow.payment.domain.Payments;
import com.ledgerflow.payment.domain.Payments.Account;
import com.ledgerflow.payment.domain.Payments.MoveResult;
import com.ledgerflow.payment.models.PaymentModel;
import com.ledgerflow.payment.models.PaymentModel.PaymentRow;
import com.ledgerflow.payment.repositories.AccountRepository;
import com.ledgerflow.payment.repositories.LedgerRepository;
import com.ledgerflow.payment.repositories.OutboxRepository;
import com.ledgerflow.payment.repositories.PaymentRepository;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;
import org.springframework.stereotype.Service;

/**
 * The two legs of the saga that run away from the request: settlement and its
 * compensating action.
 *
 * Both are called with a transaction already open - by a worker draining a
 * batch, or by the API when someone presses refund - so a leg and its outbox
 * event commit together.
 */
@Service
public class SagaService {

  private static final String CLEARING_ID = Config.SystemAccounts.CLEARING_ID;

  private final AccountRepository accounts;
  private final PaymentRepository payments;
  private final LedgerRepository ledger;
  private final OutboxRepository outbox;

  public SagaService(
      AccountRepository accounts,
      PaymentRepository payments,
      LedgerRepository ledger,
      OutboxRepository outbox) {
    this.accounts = accounts;
    this.payments = payments;
    this.ledger = ledger;
    this.outbox = outbox;
  }

  /**
   * Leg 2 of the saga: move the held funds from clearing to the receiver.
   *
   * Runs in its own transaction, some time after leg 1 - that separation is the
   * entire reason a payment can get stuck, and the reason a compensating action
   * has to exist at all.
   */
  public PaymentStatus settle(PaymentRow row) {
    if (!Payments.canSettle(row.status())) return row.status();

    if (Payments.shouldSimulateFailure(row.simulateMode(), row.attempts())) {
      return retryOrStrand(row, "SETTLEMENT_FAILED_SIMULATED");
    }

    Map<String, Account> locked = accounts.lockMany(List.of(CLEARING_ID, row.toAccountId()));
    Account clearing = locked.get(CLEARING_ID);
    Account receiver = locked.get(row.toAccountId());
    if (clearing == null || receiver == null) return retryOrStrand(row, "RECEIVER_UNAVAILABLE");

    MoveResult move = Payments.moveFunds(clearing, receiver, row.amountCents());
    if (!move.ok()) return retryOrStrand(row, move.failureReason());

    accounts.updateBalance(clearing.id(), move.fromBalanceCents());
    accounts.updateBalance(receiver.id(), move.toBalanceCents());
    ledger.postJournal(row.id(), Leg.SETTLE, move.entries());

    Instant updatedAt = payments.markCompleted(row.id(), row.attempts() + 1);
    Map<String, Object> event = PaymentModel.toEventBody(row, updatedAt);
    event.put("attempts", row.attempts() + 1);
    outbox.enqueue("payment.completed", event);
    return PaymentStatus.COMPLETED;
  }

  /**
   * A failure is not automatically fatal. Most things that break between two
   * services break briefly, so try again with a backoff and only give the money
   * back once the attempts are used up. Compensating on the first hiccup would
   * unwind perfectly good payments.
   */
  private PaymentStatus retryOrStrand(PaymentRow row, String reason) {
    int attempts = row.attempts() + 1;
    if (Payments.isExhausted(attempts)) return strand(row, reason);

    long delay = Payments.backoffMs(attempts, ThreadLocalRandom.current().nextDouble());
    Instant updatedAt = payments.scheduleRetry(row.id(), attempts, reason, delay);

    Map<String, Object> event = PaymentModel.toEventBody(row, updatedAt);
    event.put("failureReason", reason);
    event.put("attempts", attempts);
    event.put("maxAttempts", Payments.MAX_SETTLE_ATTEMPTS);
    event.put("retryInMs", delay);
    outbox.enqueue("payment.settlement_retrying", event);
    return PaymentStatus.PROCESSING;
  }

  /** Out of retries: the money stays in clearing, owed back to the sender. */
  private PaymentStatus strand(PaymentRow row, String reason) {
    Instant updatedAt =
        payments.markStranded(
            row.id(), reason, row.attempts() + 1, Config.Saga.COMPENSATE_DELAY_MS);

    Map<String, Object> event = PaymentModel.toEventBody(row, updatedAt);
    event.put("failureReason", reason);
    event.put("attempts", row.attempts() + 1);
    outbox.enqueue("payment.stuck", event);
    return PaymentStatus.AWAITING_REFUND;
  }

  public PaymentStatus compensate(PaymentRow row) {
    return compensate(row, null);
  }

  /**
   * The compensating action: return stranded funds from clearing to the sender.
   *
   * Shared by the automatic worker, the manual refund endpoint and a rejected
   * review, so all three take exactly the same path and cannot drift apart.
   */
  public PaymentStatus compensate(PaymentRow row, String reason) {
    if (!Payments.canCompensate(row.status())) return row.status();

    Map<String, Account> locked = accounts.lockMany(List.of(CLEARING_ID, row.fromAccountId()));
    Account clearing = locked.get(CLEARING_ID);
    Account sender = locked.get(row.fromAccountId());

    MoveResult move = Payments.moveFunds(clearing, sender, row.amountCents());
    if (!move.ok()) {
      // Clearing not holding the money means the ledger has been tampered with.
      // Refuse loudly rather than invent a balance.
      throw new IllegalStateException(
          "cannot compensate " + row.id() + ": clearing account " + move.failureReason());
    }

    accounts.updateBalance(clearing.id(), move.fromBalanceCents());
    accounts.updateBalance(sender.id(), move.toBalanceCents());
    ledger.postJournal(row.id(), Leg.COMPENSATE, move.entries());

    Instant updatedAt = payments.markRefunded(row.id(), reason);
    Map<String, Object> event = new LinkedHashMap<>(PaymentModel.toEventBody(row, updatedAt));
    if (reason != null) event.put("failureReason", reason);
    outbox.enqueue("payment.refunded", event);
    return PaymentStatus.REFUNDED;
  }
}
